### ISSUE TITLE
Function return type with array

### ISSUE THREAD

**Issue by deser** [microsoft/TypeScript#33158]
**TypeScript Version:**  3.6.2

**Search Terms:**
map\array\return type

**Code**

```ts
const fetchCustomVariables = (): { name?: string; type?: string }[] => {
    const variableList: { name?: string; type: string }[] = [];
    return variableList.map(v => ({ nasme: v.name, type: v.type }));
};
```
![NotificationsForm 2019-08-30 13 12 07](https://user-images.githubusercontent.com/3010698/64013435-76933500-cb28-11e9-84c4-753b6b7555fe.png)
![NotificationsForm 2019-08-30 13 14 15](https://user-images.githubusercontent.com/3010698/64013436-76933500-cb28-11e9-8e7d-690d2e543da2.png)

**Expected behavior:**
Typescript complaints that array of objects generated by `map` doesn't match function return type.

**Actual behavior:**
Typescript doesn't complaint that array of objects generated by `map` doesn't match function return type.

**Playground Link:** 
http://www.typescriptlang.org/play/#code/PTAEGUAsHsFcBsAmoDG0C2AHeBDAlgHYBQaBAzgC6gBmAphSpAMKyUYBqOATnjgEbxaZUAF5QACgCUALlABvUARzpaAflmUeBAOYBuUBQCemNRopbtoAL4BtALqiAfPKKg3qaOSoA3brwG0ADJ4lLIKSirqoJqEegbGtGYW1vaioPa6ru5c9LBcBKC+PPyCwZQAdOg4mOLeThLhOGQqst7lEbQANPEmreVGJtaSkplWmUQgoACCwpC0OSSelDT0jCxs6JzFAWQATGlSYYrKptHmsfoDpzE61vVyWW6ky0X+pSEURx1RN3FXSbEUg4xBlHqAchQ8gVGs1EqAAOQARnh3X+CN28LGRCxQA

**Related Issues:** 
No

**Comment by AnyhowStep** [comment]
```ts
variableList.map((v) : { name?: string; type: string } => (
  { nasme: v.name, type: v.type }
));
```

**Comment by AnyhowStep** [comment]
You're running into this thing called excess property checks for object literals. It isn't applied on non-object literals

**Comment by deser** [comment]
Yes I saw that, but why typescript can't guess the type? And actually, seems, it already guesses it:
![image](https://user-images.githubusercontent.com/3010698/64015565-747fa500-cb2d-11e9-9d7f-f3952f7e819b.png)

**Comment by deser** [comment]
And... `map` kind of already knows the type, it knows what array it iterates (not any). Why do I need to repeat that type for map explicitly again?

**Comment by AnyhowStep** [comment]
I knows the type. It just isn't applying the excess property checks.

**Comment by fatcerberus** [comment]
What I believe happens here is that

1. TS infers the type of the lambda expression `v => ({ ... })`.  Because there's no explicit type annotation on the lambda, it simply infers a structural type based on whatever it actually returns.  This is right because `map` allows you to output anything you want - there's no basis for it to infer any specific contract on the return type at this point.
2. The object literal has now been fully accounted for, so its "freshness" flag is removed.
3. TS asks if the function is usable with `map`.  `v` (the argument) is already the correct type as determined by contextual typing and the return type can be anything at all, so yes, it's fine.
4. Excess property checks now don't apply because the object literal(s) involved are no longer fresh. 
 It simply asks whether `{ nasme: string | undefined, type: string }[]` is structurally compatible with the return type of `fetchCustomVariables`, and it is.

If you want the excess property checks in the lambda, annotate its return type:
```ts
(v): ReturnType => value`
```

**Comment by deser** [comment]
>>Excess property checks now don't apply because the object literal(s) involved are no longer fresh.
It simply asks whether { nasme: string | undefined, type: string }[] is structurally compatible with the return type of fetchCustomVariables, and it is.

But typescript as far as I understood already infers return type of callback. It knows exactly what map will return.

**Comment by deser** [comment]
Even if types are different depending on conditions, it's quite possible to return union type... :)

**Comment by thw0rted** [comment]
It looks like #40311 fixes this.  Consider

````ts
[1,2,3].map<{a:1}>(x => ({a:1, b:2}));
````

We want the excess property `b` to be flagged.

[Current (v4.0.2) Playground](https://www.typescriptlang.org/play?#code/NoRgNATGDMC6B0BbAhgBwDwG9kC4QF8A+ACgA8ACAXkPOOzzHICMcJ8BKdgbiA) -- no error

[PR Playground](https://www.staging-typescript.org/play?ts=4.1.0-pr-40311-27#code/NoRgNATGDMC6B0BbAhgBwDwG9kC4QF8A+ACgA8ACAXkPOOzzHICMcJ8BKdgbiA) -- gives "may only specify known properties" error.  🎉 🎉  🎉
